BPOR/BPOR.Rms/Controllers/VolunteerController.cs

```diff
@@ -11,6 +12,123 @@ namespace BPOR.Rms.Controllers;
+    // For more details, see http://go.microsoft.com/fwlink/?LinkId=317598.
+    [HttpPost]
+    // [ValidateAntiForgeryToken]
+    public async Task<IActionResult> Create(
+        [Bind("FirstName,LastName,DateOfBirth,PostCode,AddressLine1,AddressLine2,AddressLine3,AddressLine4,Town,PreferredContactMethod,Email,LandLine,Mobile" +
+        ",SexRegisteredAtBirth,GenderIdentitySameAsBirth,EthnicGroup,EthnicBackground,LongTermConditionOrIllness,AreasOfResearch")]
+        VolunteerFormViewModel model, string action)
+    {
+
+        if (String.IsNullOrEmpty(model.LandLine) && String.IsNullOrEmpty(model.Mobile))
+        {
+            ModelState.AddModelError("LandLine", "At least one of either a Landline or Mobile number must be provided");
+        }
+        if (model.PreferredContactMethod == "Email" && String.IsNullOrEmpty(model.Email))
+        {
+            ModelState.AddModelError("Email", "Email must be provided when preferred contact method is email");
+        }
+
+        ValidateDateOfBirth(model.DateOfBirth);
+
+        if (!String.IsNullOrEmpty(model.Email))
+        {
+            DoesUserEmailExistInDatabase(model.Email);
+        }
+
+        if (model.DateOfBirth.HasValue && model.PostCode.HasValue && !String.IsNullOrEmpty(model.LastName))
+        {
+            DoesPostcodeSurnameDoBComboExist(model.PostCode.Value.ToString(), model.LastName, model.DateOfBirth);
+        }
+
+        if (ModelState.IsValid)
+        {
+            return RedirectToAction(nameof(VolunteerController.AccountSuccess));
+        }
+
+        return View(model);
+    }
+
+    private void DoesPostcodeSurnameDoBComboExist(string postCode, string lastName, GovUkDate dateOfBirth)
+    {
+        DateTime DoB = new DateTime(dateOfBirth.Year.Value, dateOfBirth.Month.Value, dateOfBirth.Day.Value);
+
+        var user = context.Participants.Where(p => p.LastName == lastName && 
+        p.DateOfBirth.HasValue && p.DateOfBirth.Value.Date == DoB.Date &&
+        p.Address != null &&
+        p.Address.Postcode == postCode)
+        .FirstOrDefault();
+
+        if (user != null)
+        {
+            ModelState.AddModelError("LastName", "Combination of surname, date of birth and postcode already exists and cannot be used");
+        }
+    }
+
+    private void DoesUserEmailExistInDatabase(string email)
+    {
+        var user = context.Participants.Where(p => p.Email == email).FirstOrDefault();
+
+        if (user != null)
+        {
+            ModelState.AddModelError("Email", "Email address already exists and cannot be used");
+        }
+    }
+
+    private void ValidateDateOfBirth(GovUkDate dateOfBirth)
+    {
+        if (!dateOfBirth.HasValue)
+        {
+            ModelState.AddModelError("DateOfBirth", "Date of birth is required");
+        }
+
+        if (dateOfBirth.HasValue)
+        {
+            DateOnly today = DateOnly.FromDateTime(DateTime.Today);
```

Comment:
should this be >=?
I can sign up on my 18th birthday

Reply:
The current method actually does accept a date that would on on your 18th birthday. If I input 18 09 2006, it accepts this. If I input 19 09 2006, it flags as a fail.

Reply:
ok my bad!